src/tests/test_dal_middleware.py

```diff
@@ -0,0 +1,303 @@
+
+    with dm.context():
+        assert dal.fake.authenticated_method()
+        assert middleware_requests[0] == "test_method_called"
+        with pytest.raises(SpecificException):
+            dal.fake.service.method()
+
+
+def test_dal_attribute_access_returns_dal_method_requester():
+    """
+    Verify that accessing an attribute on the DAL that does not
+    exist returns an instance of DalMethodRequester.
+
+    This is important because this starts building the call chain
+    which is used to resolve methods later.
+    """
+    dm = DataManager()
+    dal = DataAccessLayer(data_manager=dm)
+
+    # Explicitly show that an attribute that is reset returns a
+    # DMR instance
+    dal.real = "real attribute"
+    del dal.real
+    with dm.context():
+        assert isinstance(dal.real, DalMethodRequester)
+
+    with dm.context():
+        thing = dal.foo.bar.fake
+        thing2 = dal.service.foo.test.example
+        thing3 = dal.method
+        assert isinstance(thing, DalMethodRequester)
+        assert isinstance(thing2, DalMethodRequester)
+        assert isinstance(thing3, DalMethodRequester)
+        thing4 = getattr(dal, 'random')
+        assert isinstance(thing4, DalMethodRequester)
+
+    # also make sure to verify the error case where a context
+    # has not been started yet.
+    with pytest.raises(AssertionError):
+        dal.foo.bar.fake
+
+
+@pytest.mark.skip(reason="Not implemented yet, but needs to be.")
+def test_dal_getitem_attribute_access():
+    """
+    Note: We need to account for this, but I'm not sure how it will work yet.
+    Consider this test a placeholder for testing that method, which will need
+    to change, but we haven't touched yet.
+    """
+    assert False
+
+
+def test_default_dal_handler(path_arg):
+    """
+    Verify that the default dal handler performs validation on the request,
+    and then calls the dal method that was resolved.
+    """
+    called_args = []
+
+    def my_method(*args, **kwargs):
+        called_args.append((args, kwargs))
+
+    dm = DataManager()
+
+    expected_args = ['foo']
+    expected_kwargs = dict(test='bar')
+    with dm.context() as ctx:
+        request = DalMethodRequest(
+            ctx,
+            path_arg(),
+            args=expected_args,
+
+
+            kwargs=expected_kwargs
+        )
+        request.dal_method = my_method
+
+    default_handle_dal_method(request)
+
+    assert len(called_args) == 1
+    assert expected_args, expected_kwargs == called_args[0]
+
+
+@pytest.mark.xfail(reason="monkeypatching dal instance doesn't work")
+def test_dal_middleware_monkeypatch(monkeypatch, sub_service_setup):
+    """
+    Verify that pytest monkeypatching still works with dal services and methods.
+    """
+
+    def mock_user_test_method():
+        return 'mock value'
+
+    dm = sub_service_setup()
+    with dm.context() as ctx:
+        monkeypatch.setattr(ctx.dal.users, 'user_test_method', mock_user_test_method)
```

Comment:
Use:

```
        monkeypatch.setattr(
            # Mock the DAL method after anticipate is called (`func` is the original
            # function for the decorated method).
            service.__dict__[method],
            "func",
            function,
        )

```

Reply:
Yup, we talked about this elsewhere yesterday. Thank you for putting this right where we'll use it.